libraries/platforms/dromajo-vcs/bsg_manycore_platform.cpp

```diff
@@ -0,0 +1,495 @@
+  if (timeout != -1)
+  {
+    manycore_pr_err(mc, "%s: Only a timeout value of -1 is supported\n", __func__);
+    return HB_MC_INVALID;
+  }
+
+  int err;
+  int num_entries = 0;
+  do {
+    err = bp_hb_get_fifo_entries(&num_entries, type);
+  } while ((num_entries == 0) || (err != HB_MC_SUCCESS));
+
+  err = bp_hb_read_from_mc_bridge(packet, type);
+  if (err != HB_MC_SUCCESS) {
+    bsg_pr_err("Read from the %s FIFO did not succeed", typestr);
+    return HB_MC_INVALID;
+  }
+
+  return HB_MC_SUCCESS;
+}
+
+/**
+ * Read the configuration register at an index
+ * @param[in]  mc     A manycore instance initialized with hb_mc_manycore_init()
+ * @param[in]  idx    Configuration register index to access
+ * @param[out] config Configuration value at index
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_get_config_at(hb_mc_manycore_t *mc, unsigned int idx, hb_mc_config_raw_t *config) {
+  hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform);
+  hb_mc_packet_t config_req_pkt, config_resp_pkt;
+  int num_entries = 0;
+  int err;
+
+  if (idx < HB_MC_CONFIG_MAX)
+  {
+    config_req_pkt.request.x_dst = HOST_X_COORD;
+    config_req_pkt.request.y_dst = HOST_Y_COORD;
+    config_req_pkt.request.x_src = BP_HOST_LINK_X;
+    config_req_pkt.request.y_src = BP_HOST_LINK_Y;
+    config_req_pkt.request.op_v2 = HB_MC_PACKET_OP_REMOTE_LOAD;
+    config_req_pkt.request.payload = 0;
+    config_req_pkt.request.reg_id = 0;
+    config_req_pkt.request.addr = HB_MC_HOST_EPA_CONFIG_START + idx;
+
+    // Note: Potentially dangerous to write to the FIFO without checking for credits
+    // We get back credits used and not credits remaining and without the configuration
+    // there is no way to know the credits remaining.
+    err = bp_hb_write_to_mc_bridge(&config_req_pkt);
+    if (err != HB_MC_SUCCESS) {
+      bsg_pr_err("Write to the host request FIFO failed!");
+      return err;
+    }
+
+    do {
+      err = bp_hb_get_fifo_entries(&num_entries, HB_MC_FIFO_RX_RSP);
+    } while ((num_entries == 0) || (err != HB_MC_SUCCESS));
+
+    err = bp_hb_read_from_mc_bridge(&config_resp_pkt, HB_MC_FIFO_RX_RSP);
+    if (err != HB_MC_SUCCESS) {
+      bsg_pr_err("Config read failed\n");
+      return HB_MC_FAIL;
+    }
+
+    uint32_t data = config_resp_pkt.response.data;
+    *config = (data != HB_MC_HOST_OP_FINISH_CODE) ? data : 0;
+
+    return HB_MC_SUCCESS;
+  }
+
+  return HB_MC_INVALID;
+}
+
+/**
+ * Stall until the all requests (and responses) have reached their destination.
+ * @param[in]  mc     A manycore instance initialized with hb_mc_manycore_init()
+ * @param[in] timeout A timeout counter. Unused - set to -1 to wait forever.
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_fence(hb_mc_manycore_t *mc, long timeout) {
+  int credits_used, is_vacant, num_entries, err;
```

Comment:
Add a TODO (or file an issue) to fix this later. Fencing should check that the number of available credits has returned to its maximum value, not capacity.

Reply:
Yeah please file an issue and tag me in it, I'll take it on after you leave

Reply:
Not sure issue was filed

Reply:
#742